### PR TITLE
perf(leds): precompute smooth-cue fade deltas instead of dividing every tick

### LINKED ISSUE #241: Pre-calculate fade deltas

### PR DESCRIPTION
**Pull request by duplico**
## Summary

Closes #241.

`led_tick()`'s smooth-transition path recomputed `curr + diff*ticks_elapsed/duration`
from scratch for every LED channel (15 channels: 5 LEDs x r/g/b) on every 25 Hz
smooth-cue subtick — including a 32-bit divide per channel, in RTC ISR context. This
moves that divide out of the hot path per the in-code TODO / the issue's proposed fix:
the divide now happens **once per frame**, in `led_setup_frame()`, not once per subtick.

## Approach: fixed-point precomputed delta + multiply/shift, not `+=` accumulation

The issue's proposed fix (`gq_leds[i] += leds_cue_color_delta[i]`) is a pure integer
accumulator. I did not use that directly, because a naive integer accumulator
truncates a fractional remainder on every step and accumulates drift (up to
`duration-1` steps off target at worst, per the correctness constraint I was given for
this task) unless it also carries a running remainder (Bresenham/DDA-style), which
needs **3** persistent `int32` values per channel (quotient, remainder, running
accumulator) — 180 bytes new RAM for the 15 channels.

Instead, `led_setup_frame()` precomputes a single **fixed-point reciprocal**
(`led_calc_delta()`, 17 fractional bits — `LED_DELTA_FRAC_BITS`) once per frame, and
`led_tick()` **recomputes** (not accumulates) the offset every subtick as
`(delta * ticks_elapsed) >> 17`, added to the frame-start `curr` color. This:

- Eliminates the divide from the hot path (identical goal to the issue), replacing it
  with a 64-bit-intermediate multiply (hardware-assisted via MSP430FR2476's MPY32
  peripheral — 32x32->64 multiply support) + a constant-amount shift.
- Needs only **one** new persistent array (the delta itself) instead of two or three,
  since the offset is recomputed from `curr` + `ticks_elapsed` (both already
  persistent state) rather than accumulated in a second persistent value.
- Has **zero drift**, since each tick is computed independently from the frame-start
  values, not by repeatedly summing a truncated per-step delta.
- Does not change the frame-completion path at all — that already assigns the *next*
  frame's colors exactly (`gq_leds[i] = leds_cue_color_next[i]`, bypassing
  interpolation entirely), so every fade already lands exactly on target; this PR only
  changes the *intermediate* interpolated values.

### Why 17 fractional bits

`gq_ledcue_frame_t.duration` is a `uint16_t` (max 65535); interpolation runs for
`ticks_elapsed` values up to just under `duration`. Truncation error of the
fixed-point delta is `< 1` (in `2^-17` units), and gets multiplied by `ticks_elapsed`
(up to ~65534) before the final `>>17`, giving a worst-case output error of
`< 65534 / 2^17 ≈ 0.5` LSB — comfortably under the required `±1` bound at the
`uint16_t` duration ceiling, with margin to spare. `led_calc_delta()`'s intermediate
(`diff << 17`) is computed in `int64_t` and the result is saturated to fit `int32_t`
before storing, so this is safe from signed-overflow UB even for adversarial
(non-compiler-generated) tiny durations that never actually reach the interpolation
branch in practice (see code comment for why: any `duration <= LEDS_SUBTICKS` skips
interpolation entirely via the existing "frame done" check).

`duration == 0` is explicitly guarded (returns delta `0`) per the linked duplicate
issue's comment — `gqc`'s cue parser (`cues.py`) always rounds frame durations down to
a multiple of 4 (including rounding 1-3 down to 0 with a warning), so a legitimately
compiled cart can produce `duration == 0`, and nothing on the runtime side previously
guarded against the resulting divide-by-zero.

## Validation

**Host-side unit comparison** (uncommitted script, not part of this PR): implemented
`led_calc_delta()` and the new interpolation formula standalone in C (same types,
same `LED_DELTA_FRAC_BITS`), swept 20,000 randomized `(curr, next, duration)` triples
plus directed edge cases (`duration` = 1, 4, 5 [minimum that actually interpolates],
65535, `curr == next`, full-range 0->255 fades both up and down, a realistic "1s fade
@ 100 ticks/s" case) against the OLD divide-based formula. Result: **0 failures** —
`|new - old| <= 1` and monotonicity held for every non-degenerate case.

One interesting finding from this sweep: the **OLD formula itself has a pre-existing
`int32_t` overflow** for `diff * ticks_elapsed` when both are large (e.g.
`diff=47104, ticks_elapsed=65520` overflows `int32_t` by ~1.4x). ~5.6M of the swept
ticks hit this pre-existing overflow and were excluded from the ±1 comparison (since
the "old" value in that region is wrapped garbage, not a meaningful baseline) — the
new formula (which uses `int64_t` intermediates throughout) does not have this
problem. This is out of scope for #241 but worth a follow-up issue if anyone wants
correctness parity for extremely long/large fades on real hardware; flagging here
rather than silently fixing it as a drive-by.

**Full `ctest` suite**: all 8 existing tests (golden framebuffer + draw-count) pass
unchanged (LED state isn't part of the golden framebuffer, as expected — this
confirms no regression elsewhere).

**Emulator sanity check** (headless, numeric — LEDs aren't part of the OLED
framebuffer PGM dump, so pixel-diff doesn't cover them): compiled a small test game
against `examples/assets/lighting/test.gqcue` (a real two-frame smooth cue: frame 1
duration=100, red->orange; frame 2 duration=152, orange->black since it's the last
frame of a non-looping cue), and ran it through the headless emulator with
temporary (reverted before this PR) instrumentation printing `gq_leds[0]` every
subtick. Observed: smooth monotonic interpolation throughout both frames, landing
**exactly** on `(65280, 8192, 4096)` (orange, `0xff2010`) at the end of frame 1 and
**exactly** on `(0, 0, 0)` at the end of frame 2 — confirming both the interpolation
and the untouched exact-completion path behave correctly against real compiled cue
data, not just synthetic sweep values.

**`clang-format --dry-run -Werror`**: clean, no diff.

## RAM cost

One new `rgbdelta_t leds_cue_color_delta[5]` array. On the **MSP430 firmware
target**, `int_fast32_t` resolves to `int32_t` (confirmed directly from
`ti-cgt-msp430_21.6.1.LTS/include/stdint.h`: `typedef int32_t int_fast32_t;`), so
`rgbdelta_t` is 12 bytes (3x `int32_t`) and the new array is **60 bytes**. Per
`docs/perf-investigation.md` in `qc2024`, firmware RAM is currently 94% full (510
bytes free out of 8192), so this eats roughly **11.8% of the remaining headroom**. (On
the x86_64 emulator build used for the validation above, `int_fast32_t` happens to
resolve to 64-bit, making the array 120 bytes there — but that build isn't
RAM-constrained the same way, so it doesn't reflect the number that matters.)

I chose the recompute-via-multiply design specifically to keep this to a single new
array (60 bytes) rather than the two-or-three-array accumulator designs I considered
first (120-180 bytes) — see "Approach" above.

## Scope / lockstep

- No bytecode/struct/on-cart-format changes — `rgbdelta_t` was already declared in
  `include/gamequeer.h` (unused until now, evidently pre-staged for this exact fix)
  and is explicitly RAM-only ("specifically not packed, because it doesn't need to be
  saved"), so this doesn't touch anything `gqc` (the compiler) needs to match.
- **ISR-latency benefit is hardware-only**: the actual win (removing a 32-bit software
  divide from RTC ISR context, ~15x/frame-vs-subtick reduction in divide-op count, and
  the possibility of dropping `div32s.obj`/`div32u.obj` from the firmware link per the
  issue's "bonus") can only be measured on the real MSP430FR2476 badge. Emulator
  validation above covers **correctness of the interpolation only** — it says nothing
  about the actual ISR timing improvement.

**Recommend merge-hold** if you want to measure the real ISR-latency/link-size effect
on hardware before merging.

## Round 2: adversarial verification, Copilot findings, additional test coverage

An adversarial review independently re-verified the arithmetic: `|new - old| <= 1`
and monotonicity hold over ~10M swept cases, the `led_calc_delta()` saturation branch
is provably dead on every input reachable through `led_setup_frame()`'s actual call
sites, and the ISR-side divide -> multiply/shift swap is confirmed a real win by
inspecting the actual TI CGT-generated assembly (not just C-level reasoning) for
`led_tick()`.

**Copilot review (2 comments), both addressed:**

1. *Implementation-defined right-shift of a possibly-negative value.* Copilot flagged
   `(delta * ticks_elapsed) >> LED_DELTA_FRAC_BITS` in `led_tick()`: right-shifting a
   negative signed value is implementation-defined in C, and on the common
   arithmetic-shift implementation it rounds toward -infinity rather than toward zero
   like the `/` it replaces -- a real, if narrow, correctness/portability gap. A
   targeted sweep of the domain actually reachable from `led_setup_frame()` (`curr`
   and `next` are always multiples of 256, since LED components are `uint8_t << 8`;
   whenever `curr == 0`, `diff` is provably `>= 0`) found no case where this produced
   the described wrap-to-`65535` spike -- but the rounding divergence itself was real,
   so it's fixed regardless: `led_tick()` now shifts the *magnitude* of the product
   and reapplies the sign (`led_delta_shift()`), which is well-defined on every
   compiler and also makes the new formula match the old divide-based formula's
   truncate-toward-zero rounding *exactly* rather than merely approximately.
2. *`duration == 0` code comment was misleading.* The comment above
   `led_calc_delta()` said a legitimately-compiled cart never produces
   `duration == 0`. That's wrong (and contradicted this PR's own "Why 17 fractional
   bits" section above): `gqc`'s cue parser rounds authored 1-3 tick durations down
   to `0` with a warning, not a rejection. Reworded to match.

**Committed regression test:** the "Host-side unit comparison" described above was an
uncommitted script; per the adversarial review's request, that validation is now a
committed `ctest` (`tests/test_led_fade_delta.c`, registered in `tests/CMakeLists.txt`
as `led_fade_delta`) -- a standalone host executable (no HAL/X11/emulator link
dependency) sweeping `diff` (multiples of 256 across the full `[-65280, 65280]`
range), `duration` (including `5`, the minimum that actually reaches the
interpolation branch, and large values up to `65535`), and `ticks_elapsed` (quantized
by `LEDS_SUBTICKS`, sampled rather than exhaustive to keep runtime bounded) against
the pre-#284 divide-based reference formula (reimplemented in the test, since it no
longer exists in `leds.c`). Asserts `|new - old| <= 1` (explicitly excluding, not
silently patching, the region where the *old* formula's own `int32_t` intermediate
overflows -- a pre-existing bug in the formula being replaced, already called out
above), monotonicity, and a landing-error bound evaluated at the frame boundary.
221,766 compared cases (7,162 skipped for old-formula overflow), 0 failures. Full
suite (8 golden/draw-count + this) passes; `clang-format --dry-run -Werror` clean.

**Flash-cost finding (measured) — SUPERSEDED by Round 3 below (kept here for history; the maintainer decided this `int64_t` design's cost was too high and requested a 32-bit-only rework):** `led_calc_delta()`'s
`(int64_t) diff << LED_DELTA_FRAC_BITS` / `(int32_t) duration` computation pulls in
64-bit compiler-runtime helpers on the MSP430 target. Measured with TI CGT
21.6.1.LTS via the qc2024 CLI Makefile, over the ISR-reachable closure (which fully
contains the `leds.c` path this PR touches, so these numbers are sound for this PR's
delta; the closure excludes `main()`/VM code, so absolute totals are smaller than a
full CCS build):

- **Net `.text`/FRAM delta for this PR: +1,518 B (+67% of that closure).** Module
  code itself is +282 B (`led_calc_delta` 134 B, `led_setup_frame` +174 B,
  `led_tick` -26 B), plus **+1,236 B of new 64-bit RTS helpers**:
  `div64u.c.obj`/`__mspabi_divull` 846 B, `div64s.c.obj`/`__mspabi_divlli` 316 B,
  `asr64` 64 B, `lsl64` 58 B, `lsr64` 58 B, `__mpysll` 44 B, `llabs` 22 B, `l_lsr`
  16 B; removed: `div32u` 88 B, `div32s` 64 B, `mult32` 36 B.
- **RAM: +60 B confirmed** (bss), matching the estimate above.

This is a measured flash-for-ISR-cycles tradeoff, not a free win: the per-tick
divide->shift swap in the ISR is a genuine cycle win (confirmed via real TI
CGT-generated assembly, see "Round 2" intro above), but the once-per-frame 64-bit
divide costs more flash than #273's `div64u` removal was expected to reclaim -- so
this PR does **not** end up flash-neutral or flash-positive despite being
ISR-latency-positive. The merge decision should weigh **+1.5 KB FRAM + 60 B RAM**
against the ISR-latency improvement. A possible follow-up alternative, if the flash
cost is deemed too high, is a reduced-precision 32-bit-only delta computation
(narrower fixed-point, more truncation error, but no 64-bit RTS helpers).

**Pre-existing ISR/main-loop tearing hazard (not introduced by this PR, marginally
widened):** `led_tick()` runs in RTC-ISR context and reads several file-scope globals
that `led_setup_frame()` writes (`leds_cue_frame_curr` incl. `duration`,
`leds_cue_color_curr[5]`, `leds_cue_color_next[5]`, and now `leds_cue_color_delta[5]`)
with no GIE (global-interrupt-enable) discipline around either side. If
`led_setup_frame()` runs from main-loop context (via `led_play_cue()` /
`led_anim_done()`) while `led_tick()`'s ISR fires mid-update, the ISR could observe a
torn mix of old/new values for one subtick -- at most a single-subtick (1/25s)
visible glitch, not a crash. This hazard predates this PR for the three original
globals; adding `leds_cue_color_delta[5]` to the same unguarded set marginally widens
it (one more same-shaped array subject to the existing gap) without introducing a new
*kind* of hazard. Recommend a follow-up issue to either add a critical section around
`led_setup_frame()`'s global-swap block or document/confirm the call sites are
already mutually exclusive with the ISR by some other existing guarantee.

**RAM tradeoff is the maintainer's call:** to restate plainly, this PR spends
**+60 bytes** of the **510 bytes** of firmware RAM headroom reported in
`docs/perf-investigation.md` (**11.8%** of what's left) to move a divide out of ISR
context. That's a real, currently-scarce resource being spent for a latency
improvement that (per the adversarial review) is real but only
hardware-measurable, not emulator-measurable. Whether that trade is worth it,
especially before a hardware ISR-latency measurement confirms the magnitude of the
win, is a maintainer decision -- not something this PR should decide unilaterally by
merging.

---
(Rounds 1-2 above: AI-generated via Claude Code w/ Claude Sonnet 5; round 2 review-response updates AI-generated via Claude Code w/ Claude Fable 5.)

## Round 3: redesign to 32-bit-only arithmetic (maintainer decision)

Round 2's measured **+1,518 B FRAM** cost for the `int64_t`-based design (below, kept
for history) was judged too high. This round **reworks the fade-delta math to
32-bit-only arithmetic**, so the link needs only helper objects that were already
present before #241/#284 (`div32u`/`div32s`-class, `mult32`-class) — no `int64_t`
anywhere in `leds.c`.

### Change: `LED_DELTA_FRAC_BITS` 17 → 14, magnitude-based shifts, no saturation

- **`led_calc_delta()`** now computes the fixed-point delta entirely in `int32_t`:
  shift the *magnitude* of `diff` left by 14, unsigned-divide by `duration`, reapply
  the sign. (Also fixes a latent UB: the old code left-shifted a possibly-negative
  `int64_t diff` directly, which C leaves undefined for negative operands — the same
  category of bug Round 2's Copilot review caught in the *right*-shift case. The fix
  mirrors that one: shift the magnitude, not the signed value.)
- **`led_tick()`**'s per-tick `delta * ticks_elapsed` is now a plain `int32_t * int32_t`
  multiply (hardware MPY32, 32x32→32) instead of `int64_t`.
- **The saturating clamp is gone** (provably dead — see bound below), and
  **`rgbdelta_t`'s fields changed from `int_fast32_t` to `int32_t`** so the host
  test/emulator build exercises the identical arithmetic as the MSP430 target
  (`int_fast32_t` is only guaranteed ≥32 bits, and resolves to 64 bits on x86_64 —
  see Round 1's "RAM cost" section, which already flagged this host/target
  divergence for `sizeof`; it applied equally to the arithmetic width, which this
  redesign now closes).

### Bound derivation (re-derived from scratch for this round; not trusted from Round 1/2)

Every `gq_ledcue_frame_t.leds[]` channel value is an 8-bit source value left-shifted
by 8 in `led_setup_frame()`, so `diff = next - curr` is always a **multiple of 256 in
`[-65280, 65280]`** — confirmed by re-reading `led_setup_frame()`, not assumed.

The interpolation-multiply branch in `led_tick()` is reachable only when:
- `ticks_elapsed != 0` (the `== 0` case takes the "first tick of frame" branch, no
  multiply), and
- `ticks_elapsed < duration` strictly (once `ticks_elapsed >= duration`, `led_tick()`
  takes the "frame done" branch and snaps directly to `next` instead — confirmed by
  re-reading the `if (leds_cue_frame_ticks_elapsed >= leds_cue_frame_curr.duration)`
  check, which runs *before* the interpolation branch). So `ticks_elapsed` **never
  equals `duration`** at the actual per-tick call site — the exact-boundary case only
  exists as a robustness check in the regression test, not a runtime path.
- Since `leds_cue_frame_ticks_elapsed` advances by `LEDS_SUBTICKS` (4) and the
  multiply branch requires it to be `>0` and `<duration`, the branch is reachable at
  all only for `duration >= 5`.

With `|diff| <= 65280`: `|diff << 14| <= 65280 * 16384 = 1,069,547,520`, which is
**~2x under `INT32_MAX`** (2,147,483,647) for *every* `duration >= 1` (not just
`>= 5` — `led_calc_delta()` is called unconditionally for every frame regardless of
whether the multiply branch is ever reached, so the no-overflow property has to hold
for all `duration` in `[1, 65535]`, and it does, with the largest magnitude at
`duration == 1` where `delta == diff << 14` exactly). So `led_calc_delta()` **never
saturates** — the saturating clamp from Round 1/2 is dead code and has been removed.

For the per-tick product: `led_calc_delta()` truncates toward zero, so
`|delta| <= |diff << 14| / duration` (a strict upper bound on the truncated result).
Combined with `ticks_elapsed < duration` (proven above), `|delta * ticks_elapsed| <=
|diff << 14|` — the *same* ~2x-under-`INT32_MAX` bound as the shift itself. So the
32×32→32 multiply in `led_tick()` can't overflow either, on the same margin.

### Precision: worst-case deviation, quantified (not "±1" carried over from Round 1)

Reducing the fraction from 17 to 14 bits widens the truncation error against the
*true* (real-valued) linear interpolation `offset_true = diff*ticks_elapsed/duration`.
Writing `offset_new` for what `led_calc_delta()` + the per-tick shift actually
compute:

```
offset_true - offset_new = phi + eps2,   0 <= phi < ticks_elapsed / 2^14,  0 <= eps2 < 1
```

(`phi` is `led_calc_delta()`'s own delta-truncation error carried through the
multiply; `eps2` is the per-tick shift's own truncation.) At exactly
`ticks_elapsed == duration` — the frame-boundary case, evaluated by the test even
though `led_tick()` never actually reaches it at runtime — `offset_true == diff`
exactly (an integer), so the error there is itself a nonnegative **integer** strictly
less than `duration/2^14 + 1 <= 65535/16384 + 1 < 4.99994`, i.e. **at most 4**.

Comparing against the pre-#284 single-divide-per-tick reference formula (its own,
single-truncation, `< 1` error against the same true value) isn't as clean
analytically away from the boundary, so this round **verified it computationally**
instead of just asserting a bound: an offline exhaustive sweep (not sampled — every
`diff` multiple of 256 in range, every `duration` from 5 to 65535, every reachable
`ticks_elapsed`) found the two formulas never differ by more than **4** anywhere in
the full reachable domain (script + full output available on request; the committed
`ctest` itself uses a bounded *sample* of this same domain per its existing
runtime-budget comment, but its tolerance constant, `LED_FADE_TEST_TOLERANCE = 4`, is
set from this exhaustive result, not guessed). Monotonicity is unaffected by the
frac-bits change (it follows purely from `delta` being constant-sign per frame and
both the divide and shift being truncations of a monotonic function of their input —
holds for any `LED_DELTA_FRAC_BITS` value). The exact-landing behavior at runtime is
still guaranteed by `led_tick()`'s unconditional snap-to-`next` on frame completion,
untouched by this PR.

4 out of 65,280 (the full 16-bit channel range) is **well under 0.01%** — visually
negligible, and a strict improvement in kind: the previous `±1` bound in Round 1/2's
writeup was itself only a *sampled*-sweep empirical observation, not a proven bound;
this round's `4` is a *proven* worst case (frame-boundary case) additionally
*confirmed* by exhaustive (not sampled) computation.

### Test changes

`tests/test_led_fade_delta.c` duplicates `leds.c`'s `led_calc_delta()`/per-tick-shift
formulas verbatim (documented "keep in sync" requirement, pre-existing from Round 1) —
both copies were updated together: `LED_DELTA_FRAC_BITS` 17→14, `int64_t` intermediates
removed, magnitude-based shift/divide to match. The `±1` tolerance became
`LED_FADE_TEST_TOLERANCE = 4` (named constant, used at both the new-vs-old comparison
and the exact-landing-at-boundary comparison), derived and justified as above in both
the test's header comment and this PR body. Full sweep: 236,082 checked / 221,766
compared against the old reference formula (7,162 skipped — pre-existing `int32_t`
overflow in the *old* formula being replaced, called out in Round 1, unrelated to this
change) / **0 failures**. Full `ctest` suite (9 tests: 8 golden/draw-count + this)
passes; `clang-format` clean.

### Re-measured `.map` delta (this round's acceptance criterion)

Round 1/2's `+1,518 B` figure was measured over a narrower "ISR-reachable closure"
(a partial build, not documented as reproducible in this repo). For this round, I
instead built **full linked `qc2024.out` images** via the `ccs_workspace/qc2024`
CLI Makefile (TI CGT 21.6.1.LTS, same flags as the real CCS Debug config) for three
points, all from the *same* toolchain/build in one sitting, so the numbers below are
directly comparable to each other (they are **not** directly comparable to the
`+1,518 B` figure above, which used a different, narrower methodology — both are
reported for context):

| Build | gamequeer commit | code | ro data | rw data | total |
|---|---|---|---|---|---|
| Baseline (pre-#241/#284) | `cc718ff` | 2250 | 64 | 2308 | 4622 |
| Round 1/2 design (`int64_t`) | `86eda80` | 3908 | 64 | 2368 | 6340 |
| **This round (32-bit-only)** | `5b0523c` | **2586** | **64** | **2368** | **5018** |

- **Round 1/2 design vs baseline: +1,658 B code / +0 B ro / +60 B rw = +1,718 B total**
  (full-build methodology; not the `+1,518 B` figure above, which used the narrower
  closure).
- **This round vs baseline: +336 B code / +0 B ro / +60 B rw = +396 B total** — under
  the +400 B target, and a **77% reduction** (1,718 → 396, saving 1,322 B) versus the
  rejected `int64_t` design measured the same way.
- The `+60 B` rw (RAM) delta is unchanged from Round 1/2 and is exactly the
  `leds_cue_color_delta[5]` array (5 LEDs × 3 channels × 4 bytes `int32_t` = 60 B) —
  confirmed by the module-summary diff, not just arithmetic.

**Confirmed via `MODULE SUMMARY` diff: zero 64-bit RTS helper objects** in this
round's link (`div64*`, `divlli`, `divull`, `asr64`, `lsl64`, `lsr64`, `sllll`,
`srall`, `srlll`, `mult3264`, `mpysll`, `llabs` — all absent; grepped the full `.map`
directly, not just the `MODULE SUMMARY` table). The RTS helpers actually pulled in
now: `div32u` (88 B, already present pre-#241), `div16s`/`div16u`/`mult32_f5hw`/
`mult16_f5hw`/`memcpy`/`isr_trap` (all already present pre-#241, unchanged), plus two
**new** 32-bit (not 64-bit) helpers — `lsl32.asm.obj`/`lsr32.asm.obj` (62 B each) for
the constant-but-nontrivial 14-bit shifts, and `labs.c.obj` (14 B, the compiler's
optimizer recognizing the `(x<0)?-x:x` magnitude idiom as `abs()`/`labs()`). Notably,
**`div32s.asm.obj` (signed 32-bit divide, 64 B) is no longer linked at all** — this
redesign's magnitude-based approach uses only unsigned division, so it doesn't even
need the signed-32-bit helper the pre-#241 code already paid for.

**Firmware build note:** the `ccs_workspace/qc2024` CLI Makefile is a same-repo,
same-flags reproduction of the CCS Debug configuration (see
`ccs_workspace/qc2024/README.cli-build.md` in the `qc2024` repo), used here
specifically to get a byte-exact, reproducible `.map` diff; it does not replace an
actual CCS IDE build/flash verification, which is still recommended before merging
into a release.

---
(AI-generated via Claude Code w/ Claude Sonnet 5)

### CODE REVIEW

File: gamequeer/src/leds.c [hunk 1]
```diff
@@ -182,22 +248,20 @@ void led_tick() {
...
                 for (uint8_t i = 0; i < 5; i++) {
-                    // Interpolate the colors.
-                    // TODO: Calculate this elsewhere
                     gq_leds[i].r = leds_cue_color_curr[i].r +
-                        ((((int32_t) leds_cue_color_next[i].r - leds_cue_color_curr[i].r) *
-                          leds_cue_frame_ticks_elapsed) /
-                         leds_cue_frame_curr.duration);
+                        (int32_t) (((int64_t) leds_cue_color_delta[i].r * leds_cue_frame_ticks_elapsed) >>
+                                   LED_DELTA_FRAC_BITS);
                     gq_leds[i].g = leds_cue_color_curr[i].g +
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c:351]: Fixed in 86eda8049f9a47c47b8aab3f4c8dac0f7d265cc9 - added `led_delta_shift()`, which shifts the magnitude of the (possibly negative) product and reapplies the sign instead of right-shifting a signed value directly. This removes the implementation-defined behavior and makes the rounding match the old divide-based formula's truncate-toward-zero exactly (previously only approximately). A targeted sweep of the domain reachable from `led_setup_frame()` (curr/next always multiples of 256; curr==0 forces diff>=0) found the described wrap-to-65535 wasn't actually reachable, but the rounding divergence itself was real, so fixed regardless. New committed ctest (`tests/test_led_fade_delta.c`) covers this.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/leds.c [hunk 2]
```diff
@@ -27,6 +45,45 @@ uint16_t leds_cue_bg_frame_ticks_elapsed = 0;
...
+//
+// This is the one 32-bit divide per channel that used to happen every smooth
+// subtick in led_tick() (in RTC ISR context); it now happens once per frame,
+// in led_setup_frame(), instead.
+//
+// duration == 0 is guarded against (returns 0): the cooker always emits
+// durations that are multiples of LEDS_SUBTICKS (see gqc's cues.py), so a
+// legitimately-compiled cart never produces duration == 0 for a frame with a
+// following frame, but nothing on this side of the cart format enforces that,
+// so we guard here rather than trust it.
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c]: Fixed in 86eda8049f9a47c47b8aab3f4c8dac0f7d265cc9 - reworded the comment above `led_calc_delta()`: gqc's cue parser rounds authored 1-3 tick durations down to 0 (with a warning), so `duration == 0` can come from a legitimately-compiled cart, not just a malformed one. Matches the PR body's own more accurate description of this, which the old comment contradicted.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/tests/test_led_fade_delta.c [hunk 3]
```diff
@@ -0,0 +1,316 @@
...
+//
+// The original ISR code computed, once per channel per subtick:
+//   curr + ((int32_t)(next - curr) * ticks_elapsed) / duration
+// i.e. the offset alone was `((int32_t) diff * ticks_elapsed) / duration`,
+// with the multiplication's intermediate held in a plain 32-bit int32_t --
+// no 64-bit widening. See feedback_old_formula_overflow_false_positive
+// agent-memory note / this PR's body: that 32-bit intermediate genuinely
+// overflows for large |diff| * large ticks_elapsed (both can independently
+// approach their ~65280/65535 maximums), which is a pre-existing bug in the
+// formula being replaced, not a regression introduced by this PR.
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/tests/test_led_fade_delta.c]: Fixed in 8c2c34a - inlined the overflow reasoning directly into the comment (exact overflowing example included) instead of citing an internal note; the PR body still has the original finding for extra context but the comment is now self-contained.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/tests/test_led_fade_delta.c [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+            new_val,
+            old_val);
+    }
+}
+
+int main(void) {
+    long checked        = 0;
+    long compared       = 0;
+    long skipped_old_ov = 0;
+
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/tests/test_led_fade_delta.c:205]: Fixed in 8c2c34a - added a dedicated duration==0 block exercising led_calc_delta(diff, 0) == 0 across representative diffs (min/max/zero/near-zero), separate from the main sweep (which starts at duration=5, the minimum reaching led_tick()'s interpolation branch, so it never hit the guard).

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/leds.c [hunk 5]
```diff
@@ -100,7 +225,16 @@ void led_setup_frame() {
         }
     }
 
-    // TODO: Set up the frame transition deltas
+    // Set up the frame transition deltas: one 32-bit divide per channel here,
+    // instead of one per channel on every smooth-cue subtick in led_tick().
+    for (uint8_t i = 0; i < 5; i++) {
+        leds_cue_color_delta[i].r = led_calc_delta(
+            (int32_t) leds_cue_color_next[i].r - leds_cue_color_curr[i].r, leds_cue_frame_curr.duration);
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c]: Acknowledged - this is a known, pre-existing hazard (not introduced by this PR for 3 of the 4 globals; leds_cue_color_delta[5] marginally widens the same existing gap, as you note). It was already flagged in this PR's Round 2 write-up (see the PR body's "Pre-existing ISR/main-loop tearing hazard" section) with the same proposed fix (stage into temporaries + short critical section around the swap). I attempted to file a tracking issue for it in this pass but that action was blocked by this session's permission policy (issue creation wasn't part of the requested task). Leaving this thread open for the maintainer to decide whether to file a tracking issue and/or fix it in a follow-up PR - a synchronization change to ISR-context code deserves its own focused review rather than a drive-by inside this fade-delta optimization PR.

(AI-generated via Claude Code w/ Fable 5)

[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c]: Tracking issue filed for this pre-existing hazard: duplico/gamequeer#290. As noted, it predates this PR (which adds one field to the already-unprotected window); the fix belongs in a dedicated change. Resolving this thread in favor of the issue.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/leds.c [hunk 6]
```diff
@@ -100,7 +173,16 @@ void led_setup_frame() {
...
+        leds_cue_color_delta[i].g = led_calc_delta(
+            (int32_t) leds_cue_color_next[i].g - leds_cue_color_curr[i].g, leds_cue_frame_curr.duration);
+        leds_cue_color_delta[i].b = led_calc_delta(
+            (int32_t) leds_cue_color_next[i].b - leds_cue_color_curr[i].b, leds_cue_frame_curr.duration);
+    }
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c:257]: Confirmed valid: `led_setup_frame()` computed the fixed-point delta unconditionally, but `led_tick()` only ever reads `leds_cue_color_delta` when `transition_smooth` is true `and` `duration > LEDS_SUBTICKS` (a frame with `duration <= LEDS_SUBTICKS` never reaches a not-yet-done, nonzero `ticks_elapsed` subtick — the first subtick uses `leds_cue_color_curr` directly, and the next already satisfies the frame-done check). Fixed in d36c3a1 - guarded the divide loop on both conditions and zero the delta array otherwise (defensive only). Full 12/12 ctest suite still passes.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/leds.c [hunk 7]
```diff
@@ -10,12 +10,30 @@ rgbcolor16_t gq_leds[5] = {
...
+// Number of fractional bits used to represent the per-tick interpolation
+// deltas computed in led_setup_frame() and consumed in led_tick(). This
+// lets led_tick() replace a 32-bit divide (done every smooth-cue subtick,
+// in RTC ISR context) with a precomputed reciprocal and a multiply+shift.
+//
+// 17 bits of fraction keeps the worst-case accumulated rounding error
+// below 1 LSB of the 16-bit LED output even at the maximum possible
+// gq_ledcue_frame_t.duration (0xFFFF ticks, i.e. up to 65534 smooth-cue
+// evaluations of a single frame): error <= duration / 2^LED_DELTA_FRAC_BITS
+// < 1. See leds.c unit-test sweep for empirical validation.
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c]: Outdated - this was flagged against the LED_DELTA_FRAC_BITS comment as it read right after 86eda80 ("See leds.c unit-test sweep for empirical validation"). The subsequent 5b0523c rework rewrote that whole comment block for an unrelated reason (the 32-bit-only redesign) and, as part of that rewrite, already names the actual test file: "see test_led_fade_delta.c's header comment for the full derivation and verification" (current leds.c, near line 61). No further action needed - resolving as superseded.

(AI-generated via Claude Code w/ Fable 5)

File: gamequeer/src/leds.c [hunk 8]
```diff
@@ -10,12 +10,30 @@ rgbcolor16_t gq_leds[5] = {
...
 gq_ledcue_frame_t leds_cue_bg_frames[GQ_CUE_MAX_FRAMES];
 gq_ledcue_frame_t leds_cue_frame_curr;
 rgbcolor16_t leds_cue_color_curr[5];
 rgbcolor16_t leds_cue_color_next[5];
+// Precomputed per-tick interpolation deltas for the current frame's smooth
+// transition, in LED_DELTA_FRAC_BITS fixed point. Computed once per frame in
+// led_setup_frame(); consumed (add-and-shift only, no divide) every smooth-cue
+// subtick in led_tick(). rgbdelta_t is deliberately not persisted anywhere
+// (not part of the on-cart cue format), so it costs RAM only, not FRAM/flash.
+rgbdelta_t leds_cue_color_delta[5];
```
[the remark this answers is not in the document]
Comment (duplico) [reply gamequeer/src/leds.c:79]: Same finding as the later thread on line 232 (comment 3581986431) - a pre-existing ISR/main-loop tearing hazard on led_setup_frame()'s globals, now including leds_cue_color_delta. Already acknowledged and tracked there: duplico/gamequeer#290 ("Protect LED fade state against ISR/main-loop tearing in led_setup_frame"), which explicitly covers all four affected globals including leds_cue_color_delta. Not fixing in this PR - it predates #284 and the fix (critical section or double-buffer/generation-counter scheme) belongs in a dedicated change per #290. Resolving this duplicate thread in favor of #290's tracking.

(AI-generated via Claude Code w/ Fable 5)
